File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.js
const solution = (want, number, discount) => {
    return discount.reduce(
        (acc, el) => {
            if (acc.queue.length >= 10) var shifted = acc.queue.shift();
            acc.queue.push(el);

            if (want.some((e) => e === shifted)) ++number[want.indexOf(shifted)];
            if (want.some((e) => e === el)) --number[want.indexOf(el)];

            acc.res += number.every((e) => e <= 0) ? 1 : 0;

            return acc;
        },
        { queue: [], res: 0 }
    ).res;
};

File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    
    unordered_map<string, int> m;
    for (int i = 0; i < number.size(); ++i)
        m[want[i]] = number[i];

    unordered_map<string, int> bet;
    for (int i = 0; i < discount.size(); ++i)
    {
        if (i >= 10)
        {
            if (m.find(discount[i - 10]) != m.end())
                --bet[discount[i - 10]];
        }
            
        if (m.find(discount[i]) != m.end())
            ++bet[discount[i]];

        if (bet == m)
            ++answer;
    }

    return answer;
}